include/rmm/mr/device/callback_memory_resource.hpp

```diff
@@ -0,0 +1,121 @@
+/*
+ * Copyright (c) 2022, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#pragma once
+
+#include <rmm/mr/device/device_memory_resource.hpp>
+
+#include <cstddef>
+#include <functional>
+#include <utility>
+
+namespace rmm::mr {
+
+/**
+ * @brief Callback function type used by callback memory resource for allocation.
+ *
+ * The signature of the callback function is:
+ *   `void* allocate_callback_t(std::size_t bytes, cuda_stream_view stream, void* arg);
+ *
+ * * Returns a pointer to an allocation of at least `bytes` usable immediately on
+ *   `stream`. The stream-ordered behavior requirements are identical to
+ *   `device_memory_resource::allocate`.
+ *
+ * * Additional context to the callback may be passed via `arg` that is provided at
+ *   construction of the `callback_memory_resource`.
+ */
+using allocate_callback_t = std::function<void*(std::size_t, cuda_stream_view, void*)>;
+
+/**
+ * @brief Callback function type used by callback_memory_resource for deallocation.
+ *
+ * The signature of the callback function is:
+ *   `void deallocate_callback_t(void* ptr, std::size_t bytes, cuda_stream_view stream, void* arg);
+ *
+ * * Deallocates memory pointed to by `ptr`. `bytes` specifies the size of the allocation
+ *   in bytes, and must equal the value of `bytes` that was passed to the allocate callback
+ *   function. The stream-ordered behavior requirements are identical to
+ *   `device_memory_resource::deallocate`.
+ *
+ * * Additional context to the callback may be passed via `arg` that is provided at
+ *   construction of the `callback_memory_resource`.
+ */
+using deallocate_callback_t = std::function<void(void*, std::size_t, cuda_stream_view, void*)>;
+
+/**
+ * @brief A device memory resource that uses the provided callbacks for memory allocation
+ * and deallocation.
+ */
+class callback_memory_resource final : public device_memory_resource {
```

Comment:
Why do we need to terminate the hierarchy here?

Reply:
I think we do this with most of our adaptors. Mainly just don't want people inheriting from them. It also helps with devirtualization in certain cases.